lib/pengine/unpack.c

```diff
@@ -1066,53 +1065,48 @@ unpack_node_loop(xmlNode * status, bool fence, pe_working_set_t * data_set)
             continue;
         }
 
-        if (!pe__is_guest_or_remote_node(this_node)
-            && pcmk_is_set(data_set->flags, pe_flag_stonith_enabled)) {
-            // A redundant test, but preserves the order for regression tests
-            process = TRUE;
-
-        } else if (pe__is_guest_or_remote_node(this_node)) {
-            bool check = FALSE;
+        if (pe__is_guest_node(this_node)) {
             pe_resource_t *rsc = this_node->details->remote_rsc;
 
-            if(fence) {
-                check = TRUE;
-
-            } else if(rsc == NULL) {
-                /* Not ready yet */
-
-            } else if (pe__is_guest_node(this_node)
-                       && rsc->role == RSC_ROLE_STARTED
-                       && rsc->container->role == RSC_ROLE_STARTED) {
-                /* Both the connection and its containing resource need to be
-                 * known to be up before we process resources running in it.
+            if (!fence && ((rsc == NULL) || (rsc->role != RSC_ROLE_STARTED)
+                           || (rsc->container->role != RSC_ROLE_STARTED))) {
+                /* Both a guest node's connection and its containing resource
+                 * need to be known to be up before we can process the node's
+                 * resource history.
                  */
-                check = TRUE;
-
-            } else if (!pe__is_guest_node(this_node)
-                       && ((rsc->role == RSC_ROLE_STARTED)
-                           || pcmk_is_set(data_set->flags, pe_flag_shutdown_lock))) {
-                check = TRUE;
-            }
-
-            if (check) {
-                determine_remote_online_status(data_set, this_node);
-                unpack_handle_remote_attrs(this_node, state, data_set);
-                process = TRUE;
+                crm_trace("Not unpacking resource history for guest node %s "
+                          "because container and connection are not known to "
+                          "be up", id);
+                continue;
             }
 
-        } else if (this_node->details->online) {
-            process = TRUE;
+        } else if (pe__is_remote_node(this_node)) {
+            pe_resource_t *rsc = this_node->details->remote_rsc;
 
-        } else if (fence) {
-            process = TRUE;
+            if (!fence && !pcmk_is_set(data_set->flags, pe_flag_shutdown_lock)
+                && ((rsc == NULL) || (rsc->role != RSC_ROLE_STARTED))) {
```

Comment:
This behavior is slightly different from before.
- Previously, if `rsc == NULL` and `shutdown_lock` was set, then `process` was set to false and we continued to the next iteration. We never reached the `STARTED || shutdown_lock` check, because when we hit `rsc == NULL`, we skipped the rest of the `else if`s.
- Now, if `rsc == NULL` and `shutdown_lock` is set, then we don't continue to the next iteration; rather, we proceed with unpacking. Continuing to the next iteration requires that `shutdown_lock` be unset and that `rsc == NULL` or `role != STARTED`.

I don't know whether this will ever make a practical difference, and the change might have been intentional. But if I indeed worked this out correctly, then it's worth highlighting.

Reply:
You're right, that was unintentional. Looking at it more closely, I'm not sure the previous code was correct either. :-( I'll restore the previous behavior for this PR, but I'll have to do some testing with remote nodes to see if anything more should be done.